src/traffic_layout/engine.cc

```diff
@@ -566,6 +332,99 @@ RunrootEngine::copy_runroot(const std::string &original_root, const std::string
   } else {
     path_map[LAYOUT_PREFIX] = ".";
   }
+
+  YAML::Emitter yamlfile;
+  // emit key value pairs to the yaml file
+  yamlfile << YAML::BeginMap;
+  for (uint i = 0; i < dir_vector.size(); i++) {
+    yamlfile << YAML::Key << dir_vector[i];
+    yamlfile << YAML::Value << path_map[dir_vector[i]];
+  }
+  yamlfile << YAML::EndMap;
+
+  // create the file
+  std::ofstream f(Layout::relative_to(ts_runroot, "runroot.yaml"));
+  f << yamlfile.c_str();
+}
+
+// check if user want to force remove the ts_runroot
+// return true if user replies Y
+static bool
+check_remove_force()
+{
+  // check for Y/N 3 times
+  for (int i = 0; i < 3; i++) {
+    std::cout << "Are you sure to force removing runroot? (irreversible) Y/N: ";
+    std::string input;
+    std::cin >> input;
+    if (input == "Y" || input == "y")
+      return true;
+    if (input == "N" || input == "n")
+      return false;
+  }
+  ink_error("Invalid input Y/N");
+  exit(70);
+}
+
+// the function for removing the runroot
+void
+LayoutEngine::remove_runroot()
+{
+  std::string path = path_handler(arguments.get("path").value(), false, _argv[0]);
+  // check validity of the path
+  if (path.empty()) {
+    ink_fatal("Path not valid (runroot.yaml not found)");
+  }
+
+  std::string clean_root = path;
+  append_slash(clean_root);
+
+  char cwd[PATH_MAX];
+  if (getcwd(cwd, sizeof(cwd)) == nullptr) {
+    ink_fatal("unexcepted failure from getcwd() code=%d", errno);
+  }
+  std::string cur_working_dir = cwd;
+
+  if (arguments.get("force")) {
+    // the force clean
+    if (!check_remove_force()) {
+      exit(0);
+    }
+    std::cout << "Forcing removing runroot ..." << std::endl;
+    if (!remove_directory(clean_root)) {
+      ink_warning("Failed force removing runroot '%s' - %s", clean_root.c_str(), strerror(errno));
+    }
+  } else {
+    // handle the map and deleting of each directories specified in the yml file
+    RunrootMapType map = runroot_map(Layout::relative_to(clean_root, "runroot.yaml"));
+    map.erase(LAYOUT_PREFIX);
+    map.erase(LAYOUT_EXEC_PREFIX);
+    for (auto it : map) {
+      std::string dir = it.second;
+      append_slash(dir);
+      std::string later_dir = dir.substr(clean_root.size());
+      dir                   = Layout::relative_to(clean_root, later_dir.substr(0, later_dir.find_first_of("/")));
```

Comment:
You should be using `string_view` or `TextView` here, not creating substrings with `std::string`.